features/0454-present-proof-v2/README.md

```diff
@@ -0,0 +1,289 @@
+
+### States
+
+#### States for Verifier
+
+* request-sent
+* proposal-received
+* presentation-received
+* abandoned
+* done
+
+#### States for Prover
+
+* request-received
+* proposal-sent
+* presentation-sent
+* abandoned
+* done
+
+For the most part, these states map onto the transitions shown in the choreography diagram in obvious ways. However, a few subtleties are worth highlighting:
+
+* The final transitions for the verifier (Send Presentation Reject, Send Presentation Ack, and Receive Presentation Reject) all result in a final `done` state, but this `done` may or may not be the verifiers intended outcome. We may want to tweak this in a future version of the protocol.
+
+> To Do: Should the previous sentence be addressed? Removed?
+
+* A similar situation exists with the final transitions for the Prover and its final transitions (Send Presentation Reject, Receive Presentation Ack, and Receive Presentation Reject).
+* When a Prover makes a (counter-)proposal, it transitions to the `proposal-sent` state. This state is only present by implication in the choreography diagram; it essentially equates to the null or begin state in that the Prover does nothing until a presentation request arrives, triggering the leftmost transition for the Prover.
+
+Errors might occur in various places. For example, a Verifier might time out waiting for the Prover to supply a presentation. Errors trigger a `problem-report`. In this version of the protocol, all errors cause the state of both parties (the sender and the receiver of the `problem-report`) to revert to `null` (meaning it is no longer engaged in the protocol at all). Future versions of the protocol may allow more granular choices.
+
+> To Do: Should the previous sentence be addressed in this version? Should it be removed?
+
+### Choreography Diagram
+
+![present proof](credential-presentation.png)
+
+> To Do: Should the Ack or Problem Report be **required** to transition the state to "done"? This is not done in v1.0, and not done in the connection-less presentation use of the v1.0 protocol.
+
+## Messages
+
+The present proof protocol consists of these messages:
+
+* `propose-presentation` - Prover to Verifier (optional) - propose a presentation or send a counter-proposal in response to a `request-presentation` message
+* `request-presentation` - Verifier to Prover - request a presentation
+* `presentation` - Prover to Verifier - provide a presentation in response to a request
+
+In addition, the [`ack`](../0015-acks/README.md) and [`problem-report`](../0035-report-problem/README.md) messages are adopted into the protocol for confirmation and error handling.
+
+The messages that include `~attach` attachments may use either form of the embedded attachment&mdash;`base64` or `json`. In the examples below, one or the other is arbitrarily used.
```

Comment:
What's the reason for this restriction? Why is `links` excluded?

Reply:
I think this was just a missing detail, and that `links` would still be valid, too. But let's see what others say.

Reply:
I had not thought the links style would be used because of the contents of some of the messages. Anything that is hash/linked is easily shared (accidentally or on purpose) by the message receiver. That is the same for embedded data, it's just (feels) easier with a URL. Not a strong argument :-).

Note that the attachment could be embedded using JSON which in turn contains a hashlink (or hash and link). 

I'm ok with adding it if others are good with it (and OK with leaving it out).

If added here, it should also be added to the v2.0 Issue Credential, right?

Reply:
I think it should be included.

> If added here, it should also be added to the v2.0 Issue Credential, right?

Yes, for consistency.

Reply:
Yes, the two should be consistent. But we shouldn't be adding something to allow this; we should be subtracting something -- specifically, the verbiage that says we only support a subset of the attachment capability. Let's just say that we use attachments -- and if clarification is needed, we can add "in all their forms."

Reply:
> Yes, the two should be consistent. But we shouldn't be adding something to allow this; we should be subtracting something -- specifically, the verbiage that says we only support a subset of the attachment capability. Let's just say that we use attachments -- and if clarification is needed, we can add "in all their forms."

Agreed.